lib/matplotlib/image.py

```diff
@@ -162,203 +210,158 @@ def changed(self):
-            else:
-                if self._rgbacache is None:
-                    x = self.to_rgba(A, bytes=False)
-                    # Avoid side effects: to_rgba can return its argument
-                    # unchanged.
-                    if np.may_share_memory(x, A):
-                        x = x.copy()
-                    # premultiply the colors
-                    x[..., 0:3] *= x[..., 3:4]
-                    x = (x * 255).astype(np.uint8)
-                    self._rgbacache = x
-                else:
-                    x = self._rgbacache
-                im = _image.frombyte(x[yslice, xslice, :], 0)
-                if self._A.ndim == 2:
-                    im.is_grayscale = self.cmap.is_gray()
-                else:
-                    im.is_grayscale = False
-            self._imcache = im
+        if out_width_base == 0 or out_height_base == 0:
+            return None, 0, 0, None
+
+        if self.origin == 'upper':
+            # Flip the input image using a transform.  This avoids the
+            # problem with flipping the array, which results in a copy
+            # when it is converted to contiguous in the C wrapper
+            t0 = Affine2D().translate(0, -A.shape[0]).scale(1, -1)
         else:
-            im = self._imcache
+            t0 = IdentityTransform()
+
+        t0 += (
+            Affine2D()
+            .scale(
+                in_bbox.width / A.shape[1],
+                in_bbox.height / A.shape[0])
+            .translate(in_bbox.x0, in_bbox.y0)
+            + self.get_transform())
+
+        t = (t0
+            + Affine2D().translate(
+                -clipped_bbox.x0,
+                -clipped_bbox.y0)
+            .scale(magnification, magnification))
+
+        # So that the image is aligned with the edge of the axes, we want
+        # to round up the output width to the next integer.  This also
+        # means scaling the transform just slightly to account for the
+        # extra subpixel.
+        if (t.is_affine and round_to_pixel_border and
+            (out_width_base % 1.0 != 0.0 or
+             out_height_base % 1.0 != 0.0)):
+            out_width = int(ceil(out_width_base) + 1)
+            out_height = int(ceil(out_height_base) + 1)
+            extra_width = (out_width - out_width_base) / out_width_base
+            extra_height = (out_height - out_height_base) / out_height_base
+            t += Affine2D().scale(
+                1.0 + extra_width, 1.0 + extra_height)
+        else:
+            out_width = int(out_width_base)
+            out_height = int(out_height_base)
+
+        if not unsampled:
+            if len(A.shape) == 2:
+                A = self.norm(A)
+                if A.dtype.kind == 'f':
+                    # For floating-point greyscale images, we treat negative
+                    # numbers as transparent.
+                    output = np.empty((out_height, out_width), dtype=A.dtype)
+                    output[...] = -100.0
```

Comment:
Perhaps add a TODO here to update this with `np.full` at some point. Unfortunately, the numpy docs doesn't say when it was added...